{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pickle\n",
    "import os\n",
    "import sys\n",
    "import datetime\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import nltk.data\n",
    "from gensim.models import Word2Vec\n",
    "import multiprocessing\n",
    "from copy import deepcopy\n",
    "from sklearn.metrics import roc_auc_score    \n",
    "from sklearn.cross_validation import StratifiedKFold\n",
    "\n",
    "import random\n",
    "random.seed(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Load the punkt tokenizer\n",
    "tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make sure that df was built the same way as for other datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df = pickle.load(open('/mnt/mypartition/pubmed_nlp_research/C_article_replication_and_translational_classification/pickled_objects/dfJournals_trans_categories.p', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(97048, 5)\n",
      "['sentiment' 'title' 'abstract' 'qualifier_terms' 'descriptor_terms']\n"
     ]
    }
   ],
   "source": [
    "print df.shape\n",
    "print df.columns.values\n",
    "df.columns = ['label', 'title', 'abstract', 'qualifier_terms', 'descriptor_terms']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(97048, 5)\n",
      "['label' 'title' 'abstract' 'qualifier_terms' 'descriptor_terms']\n",
      "['T0' 'T1/T2' 'T3/T4']\n"
     ]
    }
   ],
   "source": [
    "print df.shape\n",
    "print df.columns.values\n",
    "print df['label'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>qualifier_terms</th>\n",
       "      <th>descriptor_terms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T0</td>\n",
       "      <td>Quantification of the relative contributions o...</td>\n",
       "      <td>Amyotrophic lateral sclerosis (ALS) and fronto...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T0</td>\n",
       "      <td>Ubiquitin Specific Protease 36 (USP36) control...</td>\n",
       "      <td>Ubiquitination of the TrkA neurotrophin recept...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T0</td>\n",
       "      <td>Functional Diversity of Cytotoxic tRNase/immun...</td>\n",
       "      <td>Contact-dependent growth inhibition (CDI) is a...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T0</td>\n",
       "      <td>Identification of a distinct substrate binding...</td>\n",
       "      <td>The type III secretion system (T3SS) effector ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T0</td>\n",
       "      <td>Role of chondroitin sulfate (CS) modification ...</td>\n",
       "      <td>Protein tyrosine phosphatase receptor type Z (...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T0</td>\n",
       "      <td>Chemically programmed bispecific antibodies in...</td>\n",
       "      <td>Chemically programmed bispecific antibodies (b...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T0</td>\n",
       "      <td>Non-mutagenic Suppression of Enterocyte Ferrop...</td>\n",
       "      <td>Iron transfer across the basolateral membrane ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T0</td>\n",
       "      <td>Structural plasticity of cholesteryl ester tra...</td>\n",
       "      <td>Cholesteryl ester transfer protein (CETP) medi...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T0</td>\n",
       "      <td>Identification of a Membrane-Bound Prepore Spe...</td>\n",
       "      <td>Pore-forming toxins (PFT) are cytolytic protei...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T0</td>\n",
       "      <td>EspR-dependent ESAT-6 secretion of Mycobacteri...</td>\n",
       "      <td>Attenuation of M. bovis BCG strain is related ...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  label                                              title  \\\n",
       "0    T0  Quantification of the relative contributions o...   \n",
       "1    T0  Ubiquitin Specific Protease 36 (USP36) control...   \n",
       "2    T0  Functional Diversity of Cytotoxic tRNase/immun...   \n",
       "3    T0  Identification of a distinct substrate binding...   \n",
       "4    T0  Role of chondroitin sulfate (CS) modification ...   \n",
       "5    T0  Chemically programmed bispecific antibodies in...   \n",
       "6    T0  Non-mutagenic Suppression of Enterocyte Ferrop...   \n",
       "7    T0  Structural plasticity of cholesteryl ester tra...   \n",
       "8    T0  Identification of a Membrane-Bound Prepore Spe...   \n",
       "9    T0  EspR-dependent ESAT-6 secretion of Mycobacteri...   \n",
       "\n",
       "                                            abstract qualifier_terms  \\\n",
       "0  Amyotrophic lateral sclerosis (ALS) and fronto...                   \n",
       "1  Ubiquitination of the TrkA neurotrophin recept...                   \n",
       "2  Contact-dependent growth inhibition (CDI) is a...                   \n",
       "3  The type III secretion system (T3SS) effector ...                   \n",
       "4  Protein tyrosine phosphatase receptor type Z (...                   \n",
       "5  Chemically programmed bispecific antibodies (b...                   \n",
       "6  Iron transfer across the basolateral membrane ...                   \n",
       "7  Cholesteryl ester transfer protein (CETP) medi...                   \n",
       "8  Pore-forming toxins (PFT) are cytolytic protei...                   \n",
       "9  Attenuation of M. bovis BCG strain is related ...                   \n",
       "\n",
       "  descriptor_terms  \n",
       "0                   \n",
       "1                   \n",
       "2                   \n",
       "3                   \n",
       "4                   \n",
       "5                   \n",
       "6                   \n",
       "7                   \n",
       "8                   \n",
       "9                   "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Apply word2vec inversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "docprob takes two lists\n",
    "* docs: a list of documents, each of which is a list of sentences\n",
    "* models: the candidate word2vec models (each potential class)\n",
    "\n",
    "it returns the array of class probabilities.  Everything is done in-memory.\n",
    "\"\"\"\n",
    "\n",
    "def docprob(docs, mods):\n",
    "    # score() takes a list [s] of sentences here; could also be a sentence generator\n",
    "    sentlist = [s for d in docs for s in d]\n",
    "    # the log likelihood of each sentence in this review under each w2v representation\n",
    "    llhd = np.array( [ m.score(sentlist, len(sentlist)) for m in mods ] )\n",
    "    # now exponentiate to get likelihoods, \n",
    "    lhd = np.exp(llhd - llhd.max(axis=0)) # subtract row max to avoid numeric overload\n",
    "    # normalize across models (stars) to get sentence-star probabilities\n",
    "    #all this transposing business does is make it so the total probability of a word \n",
    "    #   equals 1 between the 2 arrays (positive prob and negative).\n",
    "    #and the pandas data frame just puts everything into rows/columns format for easy viz\n",
    "    prob = pd.DataFrame( (lhd/lhd.sum(axis=0)).transpose() )\n",
    "    # and finally average the sentence probabilities to get the review probability\n",
    "    prob[\"doc\"] = [i for i,d in enumerate(docs) for s in d]\n",
    "    prob = prob.groupby(\"doc\").mean()\n",
    "    return prob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combined method (only has to be run once instead of 3 times)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_val = pickle.load(open('/mnt/mypartition/pubmed_nlp_research/C_article_replication_and_translational_classification/df.p', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def buildModels(train1, test1, df_val, docs_val, docs, basemodel, label):\n",
    "    \n",
    "    #add boolean labels\n",
    "    train1['boolean_label'] = 1\n",
    "    for row in train1.iterrows():\n",
    "        if row[1]['label'] != label:\n",
    "            train1.set_value(row[0], 'boolean_label', 0)\n",
    "            \n",
    "    test1['boolean_label'] = 1\n",
    "    for row in test1.iterrows():\n",
    "        if row[1]['label'] != label:\n",
    "            test1.set_value(row[0], 'boolean_label', 0)\n",
    "    \n",
    "    # ****** Split the training set into clean sentences\n",
    "    #\n",
    "    sentences_pos = []  # Initialize an empty list of sentences\n",
    "    sentences_neg = []  # Initialize an empty list of sentences\n",
    "\n",
    "    #here change to include all journal name labels of positive and negative\n",
    "    inxs_pos = np.where(train1['boolean_label'] == 1)[0].tolist()\n",
    "    inxs_neg = np.where(train1['boolean_label'] == 0)[0].tolist()\n",
    "\n",
    "    for inx in inxs_pos:\n",
    "        review = train1[\"abstract\"].iloc[inx]\n",
    "        sentences_pos += KaggleWord2VecUtility.review_to_sentences(review, tokenizer)\n",
    "\n",
    "    for inx in inxs_neg:\n",
    "        review = train1[\"abstract\"].iloc[inx]\n",
    "        sentences_neg += KaggleWord2VecUtility.review_to_sentences(review, tokenizer)\n",
    "         \n",
    "    #train models\n",
    "    models = [deepcopy(basemodel) for y in range(2)]\n",
    "    models[0].train(sentences_neg, total_examples=len(sentences_neg) )\n",
    "    models[1].train(sentences_pos, total_examples=len(sentences_pos) )\n",
    "    \n",
    "    probs = docprob(docs,models).astype(object)\n",
    "    \n",
    "    predictions = np.ones((probs.shape[0]))\n",
    "\n",
    "    predictions[np.where(probs.iloc[:,0] > 0.5)] = 0 # The first column is the negative model\n",
    "\n",
    "    score = roc_auc_score(test1[\"boolean_label\"], predictions)\n",
    "    \n",
    "    #------------------------------------------------------------------------------------------------------------    \n",
    "    print \"Testing on external validation set...\"\n",
    "\n",
    "    df_val['boolean_label'] = 1\n",
    "    for row in df_val.iterrows():\n",
    "        if row[1]['label'] != label:\n",
    "            df_val.set_value(row[0], 'boolean_label', 0)\n",
    "\n",
    "    print \"Making sentiment label predictions...\"\n",
    "    probs_val = docprob(docs_val,models).astype(object)\n",
    "\n",
    "    predictions_val = np.ones((probs_val.shape[0]))\n",
    "\n",
    "    predictions_val[np.where(probs_val.iloc[:,0] > 0.5)] = 0 # The first column is the negative model\n",
    "\n",
    "    score_val = roc_auc_score(df_val[\"boolean_label\"], predictions_val)\n",
    "\n",
    "    return score, models, score_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('ROUND', 1)\n",
      "Parsing sentences from training set\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/bs4/__init__.py:166: UserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system (\"lxml\"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.\n",
      "\n",
      "To get rid of this warning, change this:\n",
      "\n",
      " BeautifulSoup([your markup])\n",
      "\n",
      "to this:\n",
      "\n",
      " BeautifulSoup([your markup], \"lxml\")\n",
      "\n",
      "  markup_type=markup_type))\n",
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/bs4/__init__.py:207: UserWarning: \"https://clinicaltrials.gov/ct2/show/NCT02351037.\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.\n",
      "  '\"%s\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)\n",
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/bs4/__init__.py:198: UserWarning: \".\" looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.\n",
      "  '\"%s\" looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.' % markup)\n",
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/bs4/__init__.py:207: UserWarning: \"http://www.trialregister.nl/trialreg/admin/rctview.asp?TC=1173.\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.\n",
      "  '\"%s\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)\n",
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/bs4/__init__.py:207: UserWarning: \"http://www.controlled-trials.com/isrctn/pf/24165302.\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.\n",
      "  '\"%s\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building and training w2v models\n",
      "Parsing test sentences\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/bs4/__init__.py:207: UserWarning: \"http://www.fp7-improved.eu/\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.\n",
      "  '\"%s\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)\n",
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/bs4/__init__.py:207: UserWarning: \"http://clinicaltrials.gov/ct2/show/NCT00679367.\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.\n",
      "  '\"%s\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)\n",
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/bs4/__init__.py:207: UserWarning: \"http://www.clinicaltrials.gov/ct2/show/NCT00540605?term=mtn-002&rank=1.\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.\n",
      "  '\"%s\" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing validation sentences\n",
      "Scoring test set\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/nlp-vm/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.93941987441\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.810679837271\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.875031668564\n",
      "('ROUND', 2)\n",
      "Parsing sentences from training set\n",
      "Building and training w2v models\n",
      "Parsing test sentences\n",
      "Parsing validation sentences\n",
      "Scoring test set\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.941171686948\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.813878305311\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.871518711923\n",
      "('ROUND', 3)\n",
      "Parsing sentences from training set\n",
      "Building and training w2v models\n",
      "Parsing test sentences\n",
      "Parsing validation sentences\n",
      "Scoring test set\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.941924860634\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.811217740602\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.869208017249\n",
      "('ROUND', 4)\n",
      "Parsing sentences from training set\n",
      "Building and training w2v models\n",
      "Parsing test sentences\n",
      "Parsing validation sentences\n",
      "Scoring test set\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.941903398069\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.807887851547\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.871161298276\n",
      "('ROUND', 5)\n",
      "Parsing sentences from training set\n",
      "Building and training w2v models\n",
      "Parsing test sentences\n",
      "Parsing validation sentences\n",
      "Scoring test set\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.941294737074\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.814025436307\n",
      "Testing on external validation set...\n",
      "Making sentiment label predictions...\n",
      "0.873644421924\n",
      "\n",
      "---------------------------------------------------------------------------------\n",
      "TEST SET\n",
      "Average for T0 model:\n",
      "0.941142911427\n",
      "Average for T1/T2 model:\n",
      "0.811537834208\n",
      "Average for T3/T4 model:\n",
      "0.872112823587\n",
      "\n",
      "Average across all models:\n",
      "0.874931189741\n",
      "\n",
      "---------------------------------------------------------------------------------\n",
      "VALIDATION SET\n",
      "Average for T0 model:\n",
      "0.636593364198\n",
      "Average for T1/T2 model:\n",
      "0.702663512958\n",
      "Average for T3/T4 model:\n",
      "0.731006546645\n",
      "\n",
      "Average across all models:\n",
      "0.690087807933\n",
      "1:32:27\n"
     ]
    }
   ],
   "source": [
    "a = datetime.datetime.now().replace(microsecond=0)\n",
    "\n",
    "y = df['label'].values\n",
    "skf = StratifiedKFold(y, n_folds=5, shuffle=True, random_state=10)\n",
    "\n",
    "i = 1\n",
    "avgAll = []\n",
    "avgT0 = []\n",
    "avgT1T2 = []\n",
    "avgT3T4 = []\n",
    "\n",
    "avgAll_val = []\n",
    "avgT0_val = []\n",
    "avgT1T2_val = []\n",
    "avgT3T4_val = []\n",
    "\n",
    "for train_index, test_index in skf:\n",
    "\n",
    "    print(\"ROUND\", i)\n",
    "    i = i + 1\n",
    "    #use the indexes to subset the df pandas dataframe\n",
    "    train1, test1 = df.iloc[train_index], df.iloc[test_index]\n",
    "\n",
    "    print \"Parsing sentences from training set\"\n",
    "\n",
    "    # ****** Split the labeled and unlabeled training sets into clean sentences\n",
    "    #\n",
    "    sentences = []  # Initialize an empty list of sentences\n",
    "\n",
    "    for j in range(len(train1)):\n",
    "        review = train1[\"abstract\"].iloc[j]\n",
    "        sentences += KaggleWord2VecUtility.review_to_sentences(review, tokenizer)\n",
    "\n",
    "    print \"Building and training w2v models\"\n",
    "    ## create a w2v learner \n",
    "    basemodel = Word2Vec(\n",
    "        workers=multiprocessing.cpu_count(), # use your cores\n",
    "        iter=3, # iter = sweeps of SGD through the data; more is better\n",
    "        hs=1, negative=0 # we only have scoring for the hierarchical softmax setup\n",
    "        )\n",
    "    basemodel.build_vocab(sentences) \n",
    "\n",
    "    print \"Parsing test sentences\"\n",
    "    # read in the test set as a list of a list of words\n",
    "\n",
    "    docs = []\n",
    "    for k in range(len(test1)):\n",
    "        review = test1[\"abstract\"].iloc[k]\n",
    "        sents = KaggleWord2VecUtility.review_to_sentences(review, tokenizer)\n",
    "        docs.append(sents)\n",
    "        \n",
    "    print \"Parsing validation sentences\"\n",
    "    docs_val = []\n",
    "    for k in range(len(df_val)):\n",
    "        review = df_val[\"abstract\"].iloc[k]\n",
    "        sents = KaggleWord2VecUtility.review_to_sentences(review, tokenizer)\n",
    "        docs_val.append(sents)\n",
    "\n",
    "    print \"Scoring test set\"\n",
    "    tupT0 = buildModels(train1, test1, df_val, docs_val, docs, basemodel, 'T0')\n",
    "    scoreT0 = tupT0[0]\n",
    "    print scoreT0\n",
    "    avgT0.append(scoreT0)\n",
    "    avgAll.append(scoreT0)\n",
    "    modelsT0 = tupT0[1]\n",
    "    score_val = tupT0[2]\n",
    "    avgT0_val.append(score_val)\n",
    "    avgAll_val.append(score_val)\n",
    "\n",
    "    tupT1T2 = buildModels(train1, test1, df_val, docs_val, docs, basemodel, 'T1/T2')\n",
    "    scoreT1T2 = tupT1T2[0]\n",
    "    print scoreT1T2\n",
    "    avgT1T2.append(scoreT1T2)\n",
    "    avgAll.append(scoreT1T2)\n",
    "    modelsT1T2 = tupT1T2[1]\n",
    "    score_val = tupT1T2[2]\n",
    "    avgT1T2_val.append(score_val)\n",
    "    avgAll_val.append(score_val)\n",
    "\n",
    "    tupT3T4 = buildModels(train1, test1, df_val, docs_val, docs, basemodel, 'T3/T4')\n",
    "    scoreT3T4 = tupT3T4[0]\n",
    "    print scoreT3T4\n",
    "    avgT3T4.append(scoreT3T4)\n",
    "    avgAll.append(scoreT3T4)\n",
    "    modelsT3T4 = tupT3T4[1]\n",
    "    score_val = tupT3T4[2]\n",
    "    avgT3T4_val.append(score_val)\n",
    "    avgAll_val.append(score_val)\n",
    "\n",
    "print \"\"\n",
    "print \"---------------------------------------------------------------------------------\"\n",
    "print \"TEST SET\"\n",
    "print \"Average for T0 model:\"\n",
    "avgT0 = sum(avgT0) / float(len(avgT0))\n",
    "print avgT0\n",
    "print \"Average for T1/T2 model:\"\n",
    "avgT1T2 = sum(avgT1T2) / float(len(avgT1T2))\n",
    "print avgT1T2\n",
    "print \"Average for T3/T4 model:\"\n",
    "avgT3T4 = sum(avgT3T4) / float(len(avgT3T4))\n",
    "print avgT3T4\n",
    "print \"\"\n",
    "print \"Average across all models:\"\n",
    "print sum(avgAll) / float(len(avgAll))\n",
    "\n",
    "print \"\"\n",
    "print \"---------------------------------------------------------------------------------\"\n",
    "print \"VALIDATION SET\"\n",
    "print \"Average for T0 model:\"\n",
    "avgT0_val = sum(avgT0_val) / float(len(avgT0_val))\n",
    "print avgT0_val\n",
    "print \"Average for T1/T2 model:\"\n",
    "avgT1T2_val = sum(avgT1T2_val) / float(len(avgT1T2_val))\n",
    "print avgT1T2_val\n",
    "print \"Average for T3/T4 model:\"\n",
    "avgT3T4_val = sum(avgT3T4_val) / float(len(avgT3T4_val))\n",
    "print avgT3T4_val\n",
    "print \"\"\n",
    "print \"Average across all models:\"\n",
    "print sum(avgAll_val) / float(len(avgAll_val))\n",
    "    \n",
    "b = datetime.datetime.now().replace(microsecond=0)\n",
    "print(b-a)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Performance:\n",
    "Overall Average- 0.8751\n",
    "\n",
    "T0 class- 0.9411 (0.94, 0.94)\n",
    "\n",
    "T1/T2 class- 0.8122 (0.81, 0.81)\n",
    "\n",
    "T3/T4 class- 0.8719 (0.87, 0.87)\n",
    "\n",
    "*much consistency across folds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'models' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-14-5c5b86a8828d>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mpickle\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdump\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmodels\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'/mnt/mypartition/reproducing_methods/w2vmodelsJournals.p'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'wb'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'models' is not defined"
     ]
    }
   ],
   "source": [
    "pickle.dump(modelsT0, open('/mnt/mypartition/reproducing_methods/w2vmodelsJournalsT0.p', 'wb'))\n",
    "pickle.dump(modelsT1T2, open('/mnt/mypartition/reproducing_methods/w2vmodelsJournalsT1T2.p', 'wb'))\n",
    "pickle.dump(modelsT3T4, open('/mnt/mypartition/reproducing_methods/w2vmodelsJournalsT3T4.p', 'wb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### External validation on paper's dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(354, 6)\n",
      "['pmid' 'title' 'abstract' 'qualifier_terms' 'descriptor_terms' 'label']\n",
      "['T0' 'T1/T2' 'T3/T4' 'TX']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>qualifier_terms</th>\n",
       "      <th>descriptor_terms</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>23195993</td>\n",
       "      <td>Gene expression profiles in peripheral blood m...</td>\n",
       "      <td>Occupational exposure to nickel (Ni) is associ...</td>\n",
       "      <td>genetics metabolism epidemiology drug effects ...</td>\n",
       "      <td>Adult Asian Continental Ancestry Group Biomark...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25077433</td>\n",
       "      <td>SOX2 is a cancer-specific regulator of tumour ...</td>\n",
       "      <td>Although the principles that balance stem cell...</td>\n",
       "      <td>genetics metabolism pathology metabolism patho...</td>\n",
       "      <td>Animals Carcinoma, Squamous Cell Cell Line, Tu...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24107601</td>\n",
       "      <td>Imaging and cerebrospinal fluid biomarkers in ...</td>\n",
       "      <td>The pathophysiological process of Alzheimer's ...</td>\n",
       "      <td>cerebrospinal fluid genetics radionuclide imag...</td>\n",
       "      <td>Aged Aged, 80 and over Alzheimer Disease Amylo...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>24891029</td>\n",
       "      <td>Preliminary evidence of cognitive and brain ab...</td>\n",
       "      <td>To ascertain whether pediatric obesity without...</td>\n",
       "      <td>pathology physiopathology physiopathology</td>\n",
       "      <td>Adolescent Attention Brain Cognition Executive...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21691448</td>\n",
       "      <td>Obese Adolescents with Type 2 Diabetes Mellitu...</td>\n",
       "      <td>The rates of type 2 diabetes (T2DM) continue t...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>22765222</td>\n",
       "      <td>Diversity of 5S rRNA genes within individual p...</td>\n",
       "      <td>We examined intragenomic variation of paralogo...</td>\n",
       "      <td>chemistry classification genetics chemistry ge...</td>\n",
       "      <td>Bacteria DNA, Ribosomal Databases, Nucleic Aci...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>24401686</td>\n",
       "      <td>Myoinositol and glutamate complex neurometabol...</td>\n",
       "      <td>To obtain quantitative neurometabolite measure...</td>\n",
       "      <td>analogs &amp; derivatives metabolism metabolism me...</td>\n",
       "      <td>Adult Aspartic Acid Brain Injuries Case-Contro...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>22914093</td>\n",
       "      <td>Antibiotics in early life alter the murine col...</td>\n",
       "      <td>Antibiotics administered in low doses have bee...</td>\n",
       "      <td>drug effects physiology administration &amp; dosag...</td>\n",
       "      <td>Adiposity Age Factors Animals Anti-Bacterial A...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>23426830</td>\n",
       "      <td>Elevated serum anti-Müllerian hormone in adole...</td>\n",
       "      <td>Serum anti-Müllerian hormone (AMH) is linked t...</td>\n",
       "      <td>blood pathology ultrasonography blood ultrason...</td>\n",
       "      <td>Adolescent Anti-Mullerian Hormone Child Female...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>24344399</td>\n",
       "      <td>Association of obesity-mediated insulin resist...</td>\n",
       "      <td>The hypothalamus is important in hunger and me...</td>\n",
       "      <td>blood analysis anatomy &amp; histology blood blood...</td>\n",
       "      <td>Adolescent Adult Brain-Derived Neurotrophic Fa...</td>\n",
       "      <td>T0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pmid                                              title  \\\n",
       "0  23195993  Gene expression profiles in peripheral blood m...   \n",
       "1  25077433  SOX2 is a cancer-specific regulator of tumour ...   \n",
       "2  24107601  Imaging and cerebrospinal fluid biomarkers in ...   \n",
       "3  24891029  Preliminary evidence of cognitive and brain ab...   \n",
       "4  21691448  Obese Adolescents with Type 2 Diabetes Mellitu...   \n",
       "5  22765222  Diversity of 5S rRNA genes within individual p...   \n",
       "6  24401686  Myoinositol and glutamate complex neurometabol...   \n",
       "7  22914093  Antibiotics in early life alter the murine col...   \n",
       "8  23426830  Elevated serum anti-Müllerian hormone in adole...   \n",
       "9  24344399  Association of obesity-mediated insulin resist...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Occupational exposure to nickel (Ni) is associ...   \n",
       "1  Although the principles that balance stem cell...   \n",
       "2  The pathophysiological process of Alzheimer's ...   \n",
       "3  To ascertain whether pediatric obesity without...   \n",
       "4  The rates of type 2 diabetes (T2DM) continue t...   \n",
       "5  We examined intragenomic variation of paralogo...   \n",
       "6  To obtain quantitative neurometabolite measure...   \n",
       "7  Antibiotics administered in low doses have bee...   \n",
       "8  Serum anti-Müllerian hormone (AMH) is linked t...   \n",
       "9  The hypothalamus is important in hunger and me...   \n",
       "\n",
       "                                     qualifier_terms  \\\n",
       "0  genetics metabolism epidemiology drug effects ...   \n",
       "1  genetics metabolism pathology metabolism patho...   \n",
       "2  cerebrospinal fluid genetics radionuclide imag...   \n",
       "3          pathology physiopathology physiopathology   \n",
       "4                                                      \n",
       "5  chemistry classification genetics chemistry ge...   \n",
       "6  analogs & derivatives metabolism metabolism me...   \n",
       "7  drug effects physiology administration & dosag...   \n",
       "8  blood pathology ultrasonography blood ultrason...   \n",
       "9  blood analysis anatomy & histology blood blood...   \n",
       "\n",
       "                                    descriptor_terms label  \n",
       "0  Adult Asian Continental Ancestry Group Biomark...    T0  \n",
       "1  Animals Carcinoma, Squamous Cell Cell Line, Tu...    T0  \n",
       "2  Aged Aged, 80 and over Alzheimer Disease Amylo...    T0  \n",
       "3  Adolescent Attention Brain Cognition Executive...    T0  \n",
       "4                                                       T0  \n",
       "5  Bacteria DNA, Ribosomal Databases, Nucleic Aci...    T0  \n",
       "6  Adult Aspartic Acid Brain Injuries Case-Contro...    T0  \n",
       "7  Adiposity Age Factors Animals Anti-Bacterial A...    T0  \n",
       "8  Adolescent Anti-Mullerian Hormone Child Female...    T0  \n",
       "9  Adolescent Adult Brain-Derived Neurotrophic Fa...    T0  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#check that column names are the same\n",
    "print df_val.shape\n",
    "print df_val.columns.values\n",
    "print df_val['label'].unique()\n",
    "df_val[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Optional: Remove 'TX' scores to test performance\n",
    "df_val = df_val[df_val['label'] != 'TX']\n",
    "print len(df_val)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Import KaggleWord2VecUtility since didn't work from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import re\n",
    "import nltk\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from bs4 import BeautifulSoup\n",
    "from nltk.corpus import stopwords\n",
    "\n",
    "\n",
    "class KaggleWord2VecUtility(object):\n",
    "    \"\"\"KaggleWord2VecUtility is a utility class for processing raw HTML text into segments for further learning\"\"\"\n",
    "\n",
    "    @staticmethod\n",
    "    def review_to_wordlist( review, remove_stopwords=False ):\n",
    "        # Function to convert a document to a sequence of words,\n",
    "        # optionally removing stop words.  Returns a list of words.\n",
    "        #\n",
    "        # 1. Remove HTML\n",
    "        review_text = BeautifulSoup(review).get_text()\n",
    "        #\n",
    "        # 2. Remove non-letters\n",
    "        review_text = re.sub(\"[^a-zA-Z]\",\" \", review_text)\n",
    "        #\n",
    "        # 3. Convert words to lower case and split them\n",
    "        words = review_text.lower().split()\n",
    "        #\n",
    "        # 4. Optionally remove stop words (false by default)\n",
    "        if remove_stopwords:\n",
    "            stops = set(stopwords.words(\"english\"))\n",
    "            words = [w for w in words if not w in stops]\n",
    "        #\n",
    "        # 5. Return a list of words\n",
    "        return(words)\n",
    "    \n",
    "    @staticmethod\n",
    "    def review_to_words( review, remove_stopwords=False ):\n",
    "        # Function to convert a raw review to a string of words\n",
    "        # The input is a single string (a raw movie review), and \n",
    "        # the output is a single string (a preprocessed movie review)\n",
    "        #\n",
    "        # 1. Remove HTML\n",
    "        review_text = BeautifulSoup(review).get_text() \n",
    "        #\n",
    "        # 2. Remove non-letters        \n",
    "        review_text = re.sub(\"[^a-zA-Z]\", \" \", review_text) \n",
    "        #\n",
    "        # 3. Convert to lower case, split into individual words\n",
    "        words = review_text.lower().split()                             \n",
    "        #\n",
    "        # 4. Optionally remove stop words (false by default)\n",
    "        if remove_stopwords:\n",
    "            stops = set(stopwords.words(\"english\"))\n",
    "            words = [w for w in words if not w in stops]   \n",
    "        #\n",
    "        # 6. Join the words back into one string separated by space, \n",
    "        # and return the result.\n",
    "        return( \" \".join( words ))   \n",
    "\n",
    "    # Define a function to split a review into parsed sentences\n",
    "    @staticmethod\n",
    "    def review_to_sentences( review, tokenizer, remove_stopwords=False ):\n",
    "        # Function to split a review into parsed sentences. Returns a\n",
    "        # list of sentences, where each sentence is a list of words\n",
    "        #\n",
    "        # 1. Use the NLTK tokenizer to split the paragraph into sentences\n",
    "        raw_sentences = tokenizer.tokenize(review.decode('utf8').strip())\n",
    "        #\n",
    "        # 2. Loop over each sentence\n",
    "        sentences = []\n",
    "        for raw_sentence in raw_sentences:\n",
    "            # If a sentence is empty, skip it\n",
    "            if len(raw_sentence) > 0:\n",
    "                # Otherwise, call review_to_wordlist to get a list of words\n",
    "                sentences.append( KaggleWord2VecUtility.review_to_wordlist( raw_sentence, \\\n",
    "                  remove_stopwords ))\n",
    "        #\n",
    "        # Return the list of sentences (each sentence is a list of words,\n",
    "        # so this returns a list of lists\n",
    "        return sentences"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
